src/resmom/mom_pmix.c

```diff
@@ -0,0 +1,1537 @@
+ *
+ * @retval void
+ */
+void
+pbs_pmix_notify_exit(job *pjob, int exitstat, char *msg)
+{
+	pmix_status_t status;
+	pmix_info_t *pinfo;
+	size_t ninfo;
+	pmix_proc_t procname, procsrc;
+	bool flag = true;
+
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "called");
+	if (!pjob) {
+		log_event(PBSEVENT_DEBUG, 0, LOG_ERR, __func__,
+			"No job supplied, returning");
+		return;
+	}
+	snprintf(log_buffer, sizeof(log_buffer),
+		"Setting up the info array for termination");
+	log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+		pjob->ji_qs.ji_jobid, log_buffer);
+	ninfo = 3;
+	PMIX_INFO_CREATE(pinfo, ninfo);
+	/* Ensure this only goes to the job terminated event handler */
+	PMIX_INFO_LOAD(&pinfo[0], PMIX_EVENT_NON_DEFAULT, &flag, PMIX_BOOL);
+	/* Provide the exit status of the application */
+	PMIX_INFO_LOAD(&pinfo[1], PMIX_JOB_TERM_STATUS, &exitstat, PMIX_STATUS);
+	/* Provide the rank */
+	PMIX_LOAD_PROCID(&procname, pjob->ji_qs.ji_jobid, PMIX_RANK_WILDCARD);
+	PMIX_INFO_LOAD(&pinfo[2], PMIX_EVENT_AFFECTED_PROC, &procname, PMIX_PROC);
+	/* Provide the message if provided */
+	if (msg) {
+		ninfo++;
+		PMIX_INFO_LOAD(&pinfo[3], PMIX_EVENT_TEXT_MESSAGE, msg, PMIX_STRING);
+	}
+	snprintf(log_buffer, sizeof(log_buffer),
+		"Info array populated");
+	log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+		pjob->ji_qs.ji_jobid, log_buffer);
+	/*
+	 * The source of the event may not be mother superior because it
+	 * will cause the PMIx server to upcall recursively. Use an
+	 * undefined rank as the source.
+	 */
+	PMIX_LOAD_PROCID(&procsrc, pjob->ji_qs.ji_jobid, PMIX_RANK_UNDEF);
+	status = PMIx_Notify_event(PMIX_ERR_JOB_TERMINATED, &procsrc,
+		PMIX_RANGE_SESSION, pinfo, ninfo, pbs_pmix_notify_exit_cb,
+		NULL);
+	switch (status) {
+	/* The first four status cases are documented */
+	case PMIX_SUCCESS:
+		snprintf(log_buffer, sizeof(log_buffer),
+			"Exit notification pending callback");
+		log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+			pjob->ji_qs.ji_jobid, log_buffer);
+		break;
+	case PMIX_OPERATION_SUCCEEDED:
+		snprintf(log_buffer, sizeof(log_buffer),
+			"Exit notification successful");
+		log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+			pjob->ji_qs.ji_jobid, log_buffer);
+		break;
+	case PMIX_ERR_BAD_PARAM:
+		snprintf(log_buffer, sizeof(log_buffer),
+			"Exit notification contains bad parameter");
+		log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+			pjob->ji_qs.ji_jobid, log_buffer);
+		break;
+	case PMIX_ERR_NOT_SUPPORTED:
+		snprintf(log_buffer, sizeof(log_buffer),
+			"Exit notification not supported");
+		log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+			pjob->ji_qs.ji_jobid, log_buffer);
+		break;
+	default:
+		/* An undocumented error type was encountered */
+		snprintf(log_buffer, sizeof(log_buffer),
+			"Exit notification failed: %s", PMIx_Error_string(status));
+		log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+			pjob->ji_qs.ji_jobid, log_buffer);
+		break;
```

Comment:
it looks like we need to add one extra tab for all the case: statements.

Reply:
There is precedent for adding tabs and not adding them (e.g. strdup_escape()). I'll add them if you insist, but there is no rule that applies here.

Reply:
i leave the decision to you.